transaction.js

```diff
@@ -37,6 +39,10 @@ function Transaction () {
     this.data_post_start = null;
     this.data_post_delay = 0;
     this.encoding = 'utf8';
+    this.queue = {
```

Comment:
Because then *every* queue plugin needs to have redundant initialization code for `transaction.notes.queue`.  An improved version of this could move the checks out of the plugins and do them in core, and save the (very light) expense of calling each queue plugin that isn't the wanted one.

Reply:
But this is how every other plugin author has to do it.  

I have to do lots of this in my own plugins etc. to the point I have a 'common' plugin that sets all the necessary stuff in reset_transaction so that I don't have to put additional logic into core.

Thinking out loud - couldn't we add a 'helper' function into plugin.js to shorthand a lot of this as it's an annoyance of mine too.  e.g.:

`````
// UNTESTED
Plugin.prototype.get = function(o, prop) {
    var parts = prop.split('.');
    parts.shift();
    var obj = o;
    for(var i = 0; i < parts.length; i++) {
        var p = parts[i];
        if(obj[p] === undefined) {
            return obj[p]
        }
        obj = obj[p];
    }
    return obj;
}
`````

Then instead of having to do:

`if (txn.notes.foo && txn.notes.foo.bar && txn.notes.foo.bar.baz && txn.notes.foo.bar.baz == 'bop') ...`

`````
let blah = this.get(txn, 'txn.notes.queue.foo.bar.baz');
if (blah === 'bop') return next()
`````

And that would improve the situation for *everyone* and keep the logic where it should be (e.g. in the plugins themselves).

Reply:
Obviously we could also have a `this.set()` function too e.g.:

`````
this.set(txn, 'txn.notes.foo.bar.baz', 'bop');
`````

Instead of:

`````
txn.notes.foo = { 
    bar: { 
        baz: 'bop' 
    } 
}
`````

Reply:
Actually adding set/get functions like this would make it even simpler - e.g.:

`connection.transaction.notes.get('queue.wants');`
`connection.transaction.notes.set('queue.wants', 'qmail');`
`connection.notes.get('foo.bar.baz');`

Reply:
Making notes less dumb would be helpful. I won't create a standard location for queue information and it doesn't solve the Wild West nature of notes, where some other plugin might have a different idea for `notes.queue`. Creating a standard location like `txn.queue` permits plugins like elasticsearch to just save `txn.queue` and automatically get consistently meaningful queue details (wants, next_hop, result) from all queue plugins.

But a well implemented idea like setters and getters has the potential to wipe an entire class of Haraka crashing bugs away. I'd prefer putting them on the notes objects.

```js
txn.notes.set('some.wandering.path', 'value');
if (txn.notes.get('some.wandering.path') === 'value') { console.log('WE WIN') }
```

Reply:
Now that's amusing, we both arrived at the `notes.get/set`. I think that idea has some promise.

Reply:
Yeah - it appears we've had the same idea.

I get your concerns regarding the 'wild west' nature of notes but we could simply make it that all plugins that we ship (e.g. as the Haraka group) with notes like this are prefixed with '_', '_c_' or '_core_' etc. and we would document them accordingly in one place.

With setters/getters on the notes objects, we could also make a new class of 'immutable' notes that would return an error if you tried to overwrite them without the correct 'key' (where the key would be a label like the plugin name) or notes that can only be modified if the correct key is supplied with an error being returned if it is not.

Both of those ideas would also reduce the wild west nature of the notes structures and automatically identify notes clashes between plugins at run-time.

Reply:
A place to document standardized plugin shared data is the secondary reason I chose the transaction object, because docs/Transaction.md exists.

Some type of taming solution would be helpful. With NPM packaged plugins, we can no longer just grep in the repo to see every other plugins note location, and thus pick one that avoids collisions.

Reply:
Yeah - I totally get that.  

My only beef with it is simply that if you pull just Haraka down and then grep through the code looking for the logic that uses this and what else references transaction.queue.wants then you won't find anything as it's all external.  Hence my reason for not wanting to put this in core as once it's in - taking it out later becomes much harder.

We just need to address the shortcomings of the existing methods and fix/improve them as the issues don't just affect this.  Anyone using Haraka on a large scale is going to hit these issues repeatedly and I think we've come up with some good ideas that if implemented will solve a lot of them.

For this PR though - lets just put it in notes with a prefix, document it in Transaction.md and Plugins.md, merge it and then revisit this once we've implemented the getters/setters objects.

Reply:
fair enough.